p2p/security/tls/transport.go

```diff
@@ -63,7 +63,24 @@ func (t *Transport) SecureInbound(ctx context.Context, insecure net.Conn, p peer
 	for _, muxer := range t.muxers {
 		muxers = append(muxers, string(muxer))
 	}
-	// Prepend the prefered muxers list to TLS config.
+	// TLS' ALPN selection lets the server select the protocol, preferring the server's preferences.
+	// We want to prefer the client's preference though.
+	getConfigForClient := config.GetConfigForClient
+	config.GetConfigForClient = func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+	alpnLoop:
+		for _, proto := range info.SupportedProtos {
+			for _, m := range muxers {
+				if m == proto {
+					config.NextProtos = []string{proto}
```

Comment:
Don't we have to append the "libp2p" ALPN here? (It seems like were doing that below? with `	config.NextProtos = append(muxers, config.NextProtos...)`). I'm pretty sure this doesn't matter, since we know the client supports the ALPN we set here. So maybe just a comment would be fine.

Reply:
Technically we have to, but we can short-circuit it, since we know that we'll select the muxer. I'll add a comment.